libcontainer/container_linux.go

```diff
@@ -529,10 +556,37 @@ func (c *linuxContainer) newInitProcess(p *Process, cmd *exec.Cmd, messageSockPa
 		}
 	}
 	_, sharePidns := nsMaps[configs.NEWPID]
-	data, err := c.bootstrapData(c.config.Namespaces.CloneFlags(), nsMaps)
+	data, err := c.bootstrapData(c.config.Namespaces.CloneFlags(), nsMaps, initStandard)
 	if err != nil {
 		return nil, err
 	}
+
+	if c.shouldSendMountSources() {
+		mountFds := make([]string, len(c.config.Mounts))
+		for i, m := range c.config.Mounts {
+			if !m.IsBind() {
+				// StartInitialization() pairs elements on this slice with mounts.
+				// So we insert the empty string so the length of this slice and number of mounts
+				// match and they can be paired.
+				mountFds[i] = ""
+				continue
+			}
+
+			// The fd passed here will not be used: nsexec.c will overwrite it with dup3(). We just need
+			// to allocate a fd so that we know the number to pass in the environment variable. The fd
+			// must not be closed before cmd.Start(), so we reuse messageSockPair.child because the
+			// lifecycle of that fd is already taken care of.
+			cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)
+			mountFds[i] = strconv.Itoa(stdioFdCount + len(cmd.ExtraFiles) - 1)
+		}
+
+		// By using strings.Join() to create the str, we guarantee that we can parse it with strings.Split()
+		// in StartInitialization() and get a slice of the same length.
+		cmd.Env = append(cmd.Env,
+			"_LIBCONTAINER_MOUNT_FDS="+strings.Join(mountFds, ";"),
+		)
```

Comment:
Rather than defining a new microformat, can we just marshal `mountFDs` as a JSON string?

Reply:
Hmm, that is tricker to parse from C (we need it in nsexec.c). Let me give it a try tomorrow.

Reply:
> as a JSON string

Hmm, perhaps as an `[]int` rather than string. Much less conversions, probably cleaner code as well.

As for sending to nsexec, we can reuse the same netlink message as used for sending mount sources (also as an array of integers).

Reply:
Maybe we can just send both the mounts and fds as one netlink message with a basic `fd;<path>` format.

Reply:
@kolyshkin yes, used []int as we want to properly encode the numbers in json). Did it even before reading your comment :)

@cyphar hmm, I opted out of that. By the time we do the bootstrapData we don't have the fd numbers anymore nor the cmd. So, I kept the current way, using JSON

@AkihiroSuda solved, thanks! :)